### PR TITLE
[Command] Add question helper for unknown or ambiguous commands

### PR DESCRIPTION
**Pull request by hason**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #16261 |
| License | MIT |
| Doc PR |  |

t

### CODE REVIEW

**Comment by fabpot** [comment]
Without looking at the code, I cannot understand the difference between `CommandNotFoundException` and `UnknownCommandException`.

File: src/Symfony/Component/Console/Exception/AmbiguousNamespaceException.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * @author Martin Hasoň <martin.hason@gmail.com>
+ */
+class AmbiguousNamespaceException extends CommandNotFoundException
+{
+    private $namespace;
+
+    public function __construct($namespace, $alternatives = array(), $code = null, $previous = null)
+    {
+        $this->command = $namespace;
```
Comment (chalasr) [inline src/Symfony/Component/Console/Exception/AmbiguousNamespaceException.php]: typo `$this->namespace = $namespace;`

**Comment by chalasr** [comment]
👍  It would be nice to have this.

IMO having several exception classes could be avoided here. Adding an utility method to the CommandNotFoundException that formats the message based on alternatives (i.e. "Did you mean...") should be enough.

File: src/Symfony/Component/Console/Exception/UnknownCommandException.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    public function __construct($command, $alternatives = array(), $code = null, $previous = null)
+    {
+        $this->command = $command;
+
+        $message = sprintf('Command "%s" is not defined.', $command);
+
+        if ($alternatives) {
+            if (1 == count($alternatives)) {
+                $message .= "\n\nDid you mean this?\n    ";
```
Comment (ogizanagi) [inline src/Symfony/Component/Console/Exception/UnknownCommandException.php]: I'm not sure the exception should format its message itself. Instead, consumers should build the formatted message IMO.

**Comment by fabpot** [comment]
@hason Can you take comments into account and rebase? Thanks.

**Comment by fabpot** [comment]
Closing as there is no feedback. Feel free to reopen whenever it is ready.

**Pull request closed without merging**